[Merged by Bors] - Fixed bevy_ui touch input

ManevilleF opened the pull request:
# Objective

`bevy_ui` doesn't support correctly touch inputs because of two problems in the focus system:
- It attempts to retrieve touch input with a specific `0` id
- It doesn't retrieve touch positions and bases its focus solely on mouse position, absent from mobile devices

## Solution

I added a few methods to the `Touches` resource, allowing to check if **any** touch input was pressed, released or cancelled and to retrieve the *position* of the first pressed touch input and adapted the focus system.

I added a test button to the *iOS* example and it works correclty on emulator. I did not test on a real touch device as:
- Android is not working (https://github.com/bevyengine/bevy/issues/3249)
- I don't have an iOS device

ManevilleF commented:
iOS simulator:

https://user-images.githubusercontent.com/26703856/156740919-40cf4ddb-d1f4-4faa-a8c9-b12e0343d827.mp4

Obviously hovering doesn't work

ManevilleF commented:
Now that I think of it, there might be some issues with multi touch and maybe something like:

```rust
let (mouse_clicked, cursor_position) = match windows
        .get_primary()
        .and_then(|window| window.cursor_position())
    {
        None => touches_input
            .iter_just_pressed()
            .next()
            .map(|t| (true, Some(t.position())))
            .unwrap_or((false, None)),
        Some(pos) => (
            mouse_button_input.just_pressed(MouseButton::Left),
            Some(pos),
        ),
    };
```

would avoid most problems, but I think multi-touch should require a more in depth rework of the focus system

alice-i-cecile commented:
> but I think multi-touch should require a more in depth rework of the focus system

Agreed. I've run into this with leafwing_input_manager, and mutli-touch needs a lot of thoughtful design work and examination of the prior art. Important, but out of scope.

alice-i-cecile commented on crates/bevy_input/src/touch.rs:
```diff
@@ -187,6 +202,11 @@ impl Touches {
         self.just_cancelled.values()
     }
 
+    /// Retrieves the position of the first currently pressed touch
```
```suggestion
    /// Retrieves the position of the first currently pressed touch, if any
```

alice-i-cecile reviewed (approved):
A small docs nit, but the meat of the change looks good to me, and the added methods are very useful.

ManevilleF commented:
@alice-i-cecile I applied your suggestion

triscuitcircuit commented:
Works on iPhone 12 hardware. `cargo update` had to be run before the build 

https://user-images.githubusercontent.com/54428989/157711753-ba2d0245-3452-44c9-9eb2-df7a2b576547.MP4

ManevilleF commented:
@alice-i-cecile I rebased onto main, everything seems to be still working on iOS simulator. Do you think this could be reviewed soon?

@triscuitcircuit Can you run a test on your device again?

alice-i-cecile reviewed (approved):
This looks good to me, but I haven't tested it on a touch device yet.

triscuitcircuit commented:
@ManevilleF 

latest changes work

(tested on IPhone 12 hardware with IOS 15.4.1) 

https://user-images.githubusercontent.com/54428989/165460492-3d734ee1-dbaf-42fa-8a98-66a81c0124b1.MP4

alice-i-cecile commented:
Fantastic. Let me get a couple more testers and then I think this is good to go.

ManevilleF commented:
I rebased on staging and fixed the iOS example using a `Rect` instead of the new `UiRect`.

@alice-i-cecile Do you think this could be reviewed/merged soon?

Weibye commented on crates/bevy_ui/src/focus.rs:
```diff
@@ -98,7 +94,12 @@ pub fn ui_focus_system(
-        mouse_button_input.just_pressed(MouseButton::Left) || touches_input.just_pressed(0);
+        mouse_button_input.just_pressed(MouseButton::Left) || touches_input.any_just_pressed();
+
+    let cursor_position = windows
+        .get_primary()
+        .and_then(|window| window.cursor_position())
+        .or_else(|| touches_input.first_pressed_position());
```
Does this fix whatever "touch-position doesn't update"-bug that #4352 is trying to fix? I'm a bit unclear what that bug actually is, and this has `something` related to touch position so I'd figured I'd ask.

ManevilleF replied:
I don't know, #4352 seems to attempt to fix internal touch position update, so I guess not?
What I'm doing is considering basically any touch position as a cursor position

Weibye commented on examples/ios/src/lib.rs:
```diff
@@ -87,6 +89,65 @@ fn setup_scene(
         ..default()
     });
+
+    // Test ui
+    commands.spawn_bundle(UiCameraBundle::default());
```
This example isn't too large so it might not be useful but should this be split into two examples? 

First example focuses on teaching what is needed to get anything running on iOS, second example focuses on showcasing how to do UI in iOS?
2.

ManevilleF replied:
It's the only iOS example and iOS examples require quite some extra files: 
- build_rust_deps.sh
- Makefile
- its own crate files (src, Cargo.toml, etc)
- a Xcode project folder and files

So I'm not sure if adding all that again would be worth it. Not against it though

Weibye replied:
Yeah, I agree. Let's not create extra clutter until it's absolutely necessary :)

Weibye reviewed (commented):
Code looks good, neat and tidy :) Some questions:

alice-i-cecile commented:
bors r-

alice-i-cecile commented:
I *think* this is why the bors batch is failing; removing it for now so we can debug.

alice-i-cecile commented on examples/ios/src/lib.rs:
```diff
@@ -87,6 +89,65 @@ fn setup_scene(
     });
+
+    // Test ui
+    commands.spawn_bundle(UiCameraBundle::default());
```
This is causing CI to fail, and needs to be fixed.

ManevilleF replied:
Done, the example now compile and works on iOS simulator

ManevilleF commented:
@alice-i-cecile I rebased on staging and tested the example on iOS simulator, everything should be good to go

alice-i-cecile commented:
bors r+

mockersf commented:
bors seems to be timeouting

mockersf commented:
or not, sorry bors to have presumed!
